### PR TITLE
Split Transaction Collection and block Creation to Separate Task

### PR DESCRIPTION
**Pull request by bfish713**
All transaction handling should be moved to one place and out of the main consensus/da tasks

### CODE REVIEW

File: crates/hotshot/src/tasks/mod.rs [hunk 1]
```diff
@@ -407,6 +408,78 @@ where
...
+) -> TaskRunner
+where
+    CommitteeEx<TYPES, I>: ConsensusExchange<
+        TYPES,
+        Message<TYPES, I>,
+        Certificate = DACertificate<TYPES>,
+        Commitment = TYPES::BlockType,
+    >,
+{
+    // build the da task
```
Comment (shenkeyao) [inline crates/hotshot/src/tasks/mod.rs:435]: Nit: `da` -> `transaction`.

File: crates/testing/tests/network_task.rs [hunk 2]
```diff
@@ -102,6 +108,8 @@ async fn test_network_task() {
     // view number check in `publish_proposal_if_able` in consensus.rs, and we will see an error in
     // logging, but that is fine for testing as long as the network task is correctly handling
     // events.
+    output.insert(SequencingHotShotEvent::Timeout(ViewNumber::new(1)), 1);
+    output.insert(SequencingHotShotEvent::Timeout(ViewNumber::new(2)), 1);
```
Comment (shenkeyao) [inline crates/testing/tests/network_task.rs:112]: Nit: move these two lines up to line 106 (put before the comment).

File: crates/task-impls/src/transactions.rs [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+use hotshot_utils::bincode::bincode_opts;
+use snafu::Snafu;
+use std::{collections::HashSet, sync::Arc, time::Instant};
+use tracing::{debug, error, instrument, warn};
+
+#[derive(Snafu, Debug)]
+/// Error type for consensus tasks
+pub struct ConsensusTaskError {}
+
+/// Tracks state of a DA task
```
Comment (shenkeyao) [inline crates/task-impls/src/transactions.rs:36]: Nit: `DA` -> `transaction`.

File: crates/task-impls/src/transactions.rs [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+            .collect();
+        Some(txns)
+    }
+
+    /// Event filter for the transaction task
+    pub fn filter(event: &SequencingHotShotEvent<TYPES, I>) -> bool {
+        matches!(
+            event,
+            SequencingHotShotEvent::TransactionsRecv(_)
+                | SequencingHotShotEvent::DAVoteRecv(_)
```
Comment (shenkeyao) [inline crates/task-impls/src/transactions.rs:313]: I think `DAVoteRecv` should be replaced by `LeafDecided`.

**Review by shenkeyao — CHANGES_REQUESTED** [review]
I think the event filter needs to be updated. Other comments are nit issues.

File: crates/hotshot/src/tasks/mod.rs (hunk 1 above)
Comment (shenkeyao) [reply crates/hotshot/src/tasks/mod.rs:435]: Or `transactions`? Looks like we use the singular form in struct names, but the plural form in comments.

File: crates/task-impls/src/transactions.rs (hunk 4 above)
Comment (bfish713) [reply crates/task-impls/src/transactions.rs]: Yes good catch, took me a second to realize why the test were passing.  We only remove transactions on decide so we might get duplicates with this bug left in but still will be able to form blocks with transactions.  Fixed now

**Review by shenkeyao — DISMISSED** [review]

**Review by rob-maron — APPROVED** [review]

**Pull request merged**